include/ttmlir/Dialect/D2M/Utils/VirtualGrid.h

```diff
@@ -0,0 +1,92 @@
+  VirtualGridUtil(VirtualGridUtil &&) = delete;
+  VirtualGridUtil &operator=(const VirtualGridUtil &) = delete;
+  VirtualGridUtil &operator=(VirtualGridUtil &&) = delete;
+
+  /// Generates a pair of forward and inverse affine maps that allow
+  /// implementing a virtual grid as a physical-view pair of tensors/memrefs.
+  ///
+  /// The view uses the grid x shard forward map to translate pure virtual
+  /// coordinates to physical coordinates compatible with the physical grid.
+  ///
+  /// The physical memref/tensor uses the inverse map to perform core
+  /// virtualization, translating raw physical core locations at runtime into
+  /// virtual core locations that are compatible with virtual space. The inverse
+  /// map is restricted to only the grid dimensions; shard dims CANNOT
+  /// participate in virtual grid dim exprs (and vice-versa) or reblocking will
+  /// not work reliably.
+  static std::pair<mlir::AffineMap, mlir::AffineMap>
+  createCoreVirtMaps(mlir::MLIRContext *context,
+                     llvm::ArrayRef<int64_t> virtualGrid,
+                     llvm::ArrayRef<int64_t> targetGrid) {
+
+    using namespace llvm;
+    using namespace mlir;
+
+    TT_assertv(targetGrid.size() == 2ul,
+               "Target grid must have 2 dimensions {1}", targetGrid.size());
+    TT_assertv(virtualGrid.size() == 2ul,
+               "Virtual grid only supported for 2D shapes.");
+    int64_t rank = virtualGrid.size();
+
+    bool is2DWidthSharded = (rank == 2) && virtualGrid[0] == 1;
+    bool is2DHeightSharded = (rank == 2) && virtualGrid[1] == 1;
+
+    if (is2DWidthSharded || is2DHeightSharded) {
+
+      SmallVector<AffineExpr> forwardMapExprs;
+      SmallVector<AffineExpr> inverseMapExprs;
+
+      AffineExpr d0 = getAffineDimExpr(0, context);
+      AffineExpr d1 = getAffineDimExpr(1, context);
+      AffineExpr d2 = getAffineDimExpr(2, context);
+      AffineExpr d3 = getAffineDimExpr(3, context);
+      AffineExpr zero = getAffineConstantExpr(0, context);
+      AffineExpr gridRowStride = getAffineConstantExpr(targetGrid[0], context);
+      AffineExpr gridColStride = getAffineConstantExpr(targetGrid[1], context);
+
+      if (is2DWidthSharded) {
+        forwardMapExprs = {d1.floorDiv(gridColStride), d1 % gridColStride, d2,
+                           d3};
+        // note: inverse map results are (deviceIndex, gridY, gridX)
+        inverseMapExprs = {zero, zero, d0 * gridRowStride + d1};
+      } else if (is2DHeightSharded) {
+        forwardMapExprs = {d0 % gridRowStride, d0.floorDiv(gridRowStride), d2,
+                           d3};
+        inverseMapExprs = {zero, d1 * gridColStride + d0, zero};
+      }
+      auto forward =
+          mlir::AffineMap::get(2 * rank, 0, forwardMapExprs, context);
+      auto inverse = mlir::AffineMap::get(rank, 0, inverseMapExprs, context);
+      return {forward, inverse};
+    } else {
+      TT_assertv((is2DWidthSharded || is2DHeightSharded),
```

Comment:
:warning: **llvm\-else\-after\-return** :warning:
do not use `` else `` after `` return ``
```suggestion
    }       TT_assertv((is2DWidthSharded || is2DHeightSharded),
```

Reply:
Fixed.